### DISCUSSION TITLE
Electromagnetic fields (EEMF_macros.mortran) in DOSRZnrc

### DISCUSSION THREAD

**Discussion by SimulatedWorld2** [nrc-cnrc/EGSnrc#975] [Questions] [answered]
I am trying to simulate PDDs with/without magnetic fields. I was thinking of using dosrznrc. Is it possible to use `EEMF_macros.mortran` with dosrznrc (similar to egs_chamber/cavity)? If so, how can I enable it? I tried to include it in `dosrznrc.make` and also in makefile, but it doesn't work. Thanks in advance!

#### Example `dosrznrc.make`

```
SOURCES = \
	$(EGS_SOURCEDIR)egsnrc.macros \
	$(EGS_UTILS)timing.macros\
	$(MACHINE_MACROS) \
	$(RANDOM).macros\
	$(EGS_SOURCEDIR)transportp.macros\
        $(EGS_SOURCEDIR)pegs4_macros.mortran\
	$(EGS_UTILS)phsp_macros.mortran $(IAEA_PHSP_MACROS)\
	$(EGS_UTILS)srcrz.macros\
	**$(EGS_SOURCEDIR)EEMF_macros.mortran**  \
	$(USER_CODE).mortran \
	$(RANDOM).mortran \
	$(EGS_UTILS)srcrz.mortran\
	$(EGS_UTILS)ensrc.mortran\
	$(EGS_UTILS)geomrz.mortran\
	$(EGS_SOURCEDIR)get_inputs.mortran\
        $(EGS_SOURCEDIR)get_media_inputs.mortran\
	$(EGS_UTILS)grids.mortran\
	$(EGS_UTILS)nrcaux.mortran \
	$(EGS_UTILS)xvgrplot.mortran \
	$(MACHINE_MORTRAN) \
	$(EGS_SOURCEDIR)egs_utilities.mortran \
	$(EGS_SOURCEDIR)egs_parallel.mortran \
        $(EGS_SOURCEDIR)pegs4_routines.mortran\
	$(EGS_SOURCEDIR)egsnrc.mortran	\
```

**Comment by ftessier** [comment, marked as answer]
Yes, it is possible to use EEMF in DOSRZnrc, since it is implemented in the mortran backend. You need to define the EMF parameters in your input file, as detailed in the [instructions](https://physics.carleton.ca/sites/default/files/clrp/databases/EMFmacros/EMFMacrosInEGSnrcReport_toNRC_Jan13_2017.pdf), section 4. If it does not work as expected, please share a small sample input file so that we can take a look at it.

**Reply by SimulatedWorld2** [reply]
Thanks for the quick reply! Do I understand it correctly that I don't need to change anything in the Makefile/DOSRZnrc.make-files? Anyway, without change the results are exactly the same with/without magnetic field, with the above change they are very similar. With a monoenergy of 4MeV and B=1.5T in the x-direction, I expect a 2%-ish shift of the PDD to the surface (z-direction), which is obviously not the case. Is there a better way to check if the magnetic field works?

I use the input syntax described in your attached document. I also use this input for egs_chamber/cavity, where it works. It looks like this: 

```
:start MC transport parameter:
    magnetic field = 1.5 , 0, 0 # magnetic field in Tesla
    EM ESTEPE = 0.02
    [...]
:stop MC transport parameter:
```

The `.egsinp` file I used is attached. It is essentially the depth-dose example in the "example" folder.

[depth_dose_example.txt](https://github.com/nrc-cnrc/EGSnrc/files/10982013/depth_dose_example.txt)

**Reply by ftessier** [reply]
You still need to add `$(EGS_SOURCEDIR)EEMF_macros.mortran` in `dosrznrc.make`. One way to check what is happening is to rebuild the geometry with `egs++` and use, say, the egs_chamber application to generate particle tracks. Then it is possible to assess the curving of the tracks due to the magnetic field in the viewer!

**Reply by rtownson** [reply]
To use the enhanced magnetic fields, yes you have to add the EEMF line that you suggested in the Makefile, and recompile. Watch for the output from the compiler:
```
  -> appending /home/rwtownson/EGSnrc/HEN_HOUSE/src/EEMF_macros.mortran
```

When running the simulation, watch for the output:
```
 Magnetic Field [T]                                               1.50      0.00      0.00
 EM ESTEPE                                                        0.02
```
which I do see when running your input file. That means the magnetic field is being applied. I get different results:

```
With magnetic field
         0.0000              1000.0000
    0.0000-----------------------
          |IRL  2 IZ   1 IX   1 |
          |T  : 1.033E-11+- 4.4%|
          |T-S: 9.663E-12+- 4.4%|
    2.0000-----------------------
          |IRL  3 IZ   2 IX   1 |
          |T  : 1.370E-11+- 3.7%|
          |T-S: 1.281E-11+- 3.8%|
    4.0000-----------------------
          |IRL  4 IZ   3 IX   1 |
          |T  : 1.323E-11+- 3.8%|
          |T-S: 1.234E-11+- 3.8%|

Without magnetic field
         0.0000              1000.0000
    0.0000-----------------------
          |IRL  2 IZ   1 IX   1 |
          |T  : 1.010E-11+- 4.2%|
          |T-S: 9.463E-12+- 4.3%|
    2.0000-----------------------
          |IRL  3 IZ   2 IX   1 |
          |T  : 1.336E-11+- 3.6%|
          |T-S: 1.251E-11+- 3.6%|
    4.0000-----------------------
          |IRL  4 IZ   3 IX   1 |
          |T  : 1.270E-11+- 3.7%|
          |T-S: 1.188E-11+- 3.7%|
          
```

**Reply by SimulatedWorld2** [reply]
Thanks for the clarification, I was a little confused!
I used a higher energy (20MeV) and there I can see a clear shift, the magnetic field definitely works!

Thanks for the quick replies and help!

**Comment by rtownson** [comment]
Make sure that the magnetic field numbers are separated by commas, not just spaces! This is a requirement for the Mortran codes but not in egs++, so is a very easy mistake to make.
